src/internet_identity/src/storage/tests.rs

```diff
@@ -0,0 +1,175 @@
+    storage.flush();
+
+    let mut buf = vec![0; HEADER_SIZE];
+    memory.read(0, &mut buf);
+    assert_eq!(buf, hex::decode("494943010600000087d6120000000000404b4c000000000000084343434343434343434343434343434343434343434343434343434343434343").unwrap());
+}
+
+#[test]
+fn should_serialize_first_record() {
+    let memory = VectorMemory::default();
+    let mut storage =
+        Storage::<Vec<DeviceDataInternal>, VectorMemory>::new((123, 456), memory.clone());
+    let user_number = storage.allocate_user_number().unwrap();
+    assert_eq!(user_number, 123u64);
+
+    let device_vec = sample_anchor_record();
+    storage.write(user_number, device_vec.clone()).unwrap();
+
+    let mut buf = [0u8; 192];
+    memory.read(512, &mut buf);
+    let decoded_from_memory: Vec<DeviceDataInternal> = candid::decode_one(&buf[2..]).unwrap();
+    assert_eq!(decoded_from_memory, device_vec);
+}
+
+#[test]
+fn should_serialize_subsequent_record_to_expected_memory_location() {
+    const EXPECTED_RECORD_OFFSET: u64 = 204_800;
+    let memory = VectorMemory::default();
+    let mut storage =
+        Storage::<Vec<DeviceDataInternal>, VectorMemory>::new((123, 456), memory.clone());
+    for _ in 0..100 {
+        storage.allocate_user_number().unwrap();
+    }
+    let user_number = storage.allocate_user_number().unwrap();
+    assert_eq!(user_number, 223u64);
+
+    let device_vec = sample_anchor_record();
+    storage.write(user_number, device_vec.clone()).unwrap();
+
+    let mut buf = [0u8; 192];
+    memory.read(512 + EXPECTED_RECORD_OFFSET, &mut buf);
+    let decoded_from_memory: Vec<DeviceDataInternal> = candid::decode_one(&buf[2..]).unwrap();
+    assert_eq!(decoded_from_memory, device_vec);
+}
+
+#[test]
+fn should_not_write_using_anchor_number_outside_allocated_range() {
+    let memory = VectorMemory::default();
+    let mut storage =
+        Storage::<Vec<DeviceDataInternal>, VectorMemory>::new((123, 456), memory.clone());
+    storage.allocate_user_number().unwrap();
+
+    let result = storage.write(222, sample_anchor_record().clone());
+    assert!(matches!(result, Err(StorageError::BadUserNumber(_))))
+}
+
+#[test]
+fn should_deserialize_first_record() {
+    let memory = VectorMemory::default();
+    memory.grow(1);
+    let mut storage =
+        Storage::<Vec<DeviceDataInternal>, VectorMemory>::new((123, 456), memory.clone());
+    let user_number = storage.allocate_user_number().unwrap();
+    assert_eq!(user_number, 123u64);
+
+    let device_vec = sample_anchor_record();
+    let buf = candid::encode_one(&device_vec).unwrap();
+    memory.write(512, &(buf.len() as u16).to_le_bytes());
+    memory.write(514, &buf);
+
+    let read_from_storage = storage.read(123).unwrap();
+    assert_eq!(read_from_storage, device_vec);
+}
+
+#[test]
+fn should_deserialize_subsequent_record_at_expected_memory_location() {
+    const EXPECTED_RECORD_OFFSET: u64 = 204_800;
+    let memory = VectorMemory::default();
+    memory.grow(4);
```

Comment:
I'm not super familiar with the implementation. What happens here?

Reply:
This grows the stable memory to fit the address `512 + EXPECTED_RECORD_OFFSET` such that we don't get an out of bounds error on `memory.write`. Memory has to be allocated explicitly in 64kb chunks (called pages). And this address happens to lie in chunk 4.